lib/phoenix_html/link.ex

```diff
@@ -133,21 +133,44 @@ defmodule Phoenix.HTML.Link do
 
   All other options are forwarded to the underlying button input.
   """
-  def button(text, opts) do
-    {to, opts} = Keyword.pop(opts, :to)
-    {method, opts} = Keyword.pop(opts, :method, :post)
 
-    {form, opts} = form_options(opts, method, "button")
+  def button(opts) do
+    {to, form, opts} = extract_button_options(opts)
 
-    unless to do
-      raise ArgumentError, "option :to is required in button/2"
+    {do_option, opts} = pop_required_option(opts, :do, "option :do is required in button/1")
+
+    form_tag(to, form) do
+      Phoenix.HTML.Form.submit(opts, [do: {:safe, do_option}])
     end
+  end
+
+  def button(text, opts) do
+    {to, form, opts} = extract_button_options(opts)
 
     form_tag(to, form) do
       Phoenix.HTML.Form.submit(text, opts)
     end
   end
 
+  defp extract_button_options(opts) do
+    {to, opts} = pop_required_option(opts, :to, "option :to is required in button/2")
+    {method, opts} = Keyword.pop(opts, :method, :post)
+
+    {form, opts} = form_options(opts, method, "button")
+
+    {to, form, opts}
+  end
+
+  defp pop_required_option(opts, key, error_message) do
```

Comment:
Good idea. I would probably add a `!` to the end of the function name. Also, in the `link` function above, in the same file, there are at least 2 occasions we could use this too. Can you please refactor that? Thank you!